core/src/main/java/smile/classification/DecisionTree.java

```diff
@@ -654,20 +704,26 @@ public Node findBestSplit(int n, int low, int high, int[] count, int[] falseCoun
                     if (gain > splitNode.splitScore) {
                         // new best split
                         splitNode.splitFeature = j;
-                        splitNode.splitValue = (x[o][j] + prevx) / 2;
+                        splitNode.splitValue = (x_oj + prevx) / 2;
                         splitNode.splitScore = gain;
                         splitNode.trueChildOutput = Math.whichMax(trueCount);
                         splitNode.falseChildOutput = Math.whichMax(falseCount);
                     }
 
-                    prevx = x[o][j];
-                    prevy = y[o];
-                    trueCount[y[o]] += samples[o];
+                    prevx = x_oj;
+                    prevy = y_o;
+                    trueCount[y_o] += samples[o];
                 }
             } else {
                 throw new IllegalStateException("Unsupported attribute type: " + attributes[j].getType());
             }
 
+            if(replaceCount + (countNaN > 0 ? 1: 0) <= 1) {
+                synchronized (constFeatures) {
```

Comment:
This synchronized block won't properly protect the modifications to constFeatures: two threads could replace it with different versions. To set this up, I think you need three threads...
Say thread A is running with j=1, B is running with j=2, and C is running with j=3. Initially, this.constFeatures = (say) CF1.
A sees that feature 1 is constant, and does
synchronized(CF1) { this.constFeatures = CF2; } where CF2 = CF1 + "1"
Some time later, B sees that feature 2 is constant - and after A did the above, it does (this is one possible legal execution)
syncronized(CF2) { this.constFeatures = CF3; } where CF3 = CF2 + "2".
Then thread C sees that feature 3 is constant, and does (this is also a legal execution)
synchronized(CF1) { this.constFeatures = CF4; } where CF4 = CF2 + "3". "2" has been lost.
Why did C do synchronized(CF1), not synchronized(CF2) or synchronized(CF3), and why did C base CF4 off CF2? Basically, until they acquire a lock, B and C are under no obligation to witness any modifications A has made - they might or they might not. So B happens to witness them, so it does synchronized(CF2) and modifies this.constFeatures to point to CF3. But C doesn't witness A's changes, or B's changes, and does syncronized(CF1). After it acquires the CF1 lock, it is then required to see modifications that were made while holding the lock - which were thread A's modifications of this.constFeatures to point to CF2. C isn't required to see B's modifications because those were made while holding a different lock (CF2)... so CF4 is based off CF2, and B's modification is lost.

I think that "synchronized(this)" is more reliable easier to understand, since the lock being used doesn't shift during execution.

Reply:
> This synchronized block won't properly protect the modifications to constFeatures
It's intended.

No need to be atomic but synchronize `this.constFeature` in CPU caches when reading/writing `this.constFeature`.

Reply:
The real issue is that you're synchonizing on the object currently stored in this.constFeatures - which changes over time. To get consistent behaviour, you should synchronize on something that doesn't change over time; "this" is handy :-)

Reply:
synchronizing on `this.constFeatures` is enough since I'm using `synchronized(this.constFeatures)` for memory fence. Synchronizing changes by a particular thread to another threads. Using stale constFeatures does not have side effects.

Reply:
Synchronozing on this.constFeatures
- is harder to understand than just synchronizing on "this" because this.constFeatures' value can change, so the reader has to do a lot of logic to figure out the possible effects
- does not cover all cases (see my example above where a const feature is forgotten about)
Synchronization isn't a global memory barrier, and doesn't necessarily affect CPU caches at all (if a CPU core knows that the cache line holding the lock is not present in any other core's cache, say via https://en.wikipedia.org/wiki/MOESI_protocol, then the lock acquisition is done purely locally to that one core's cache; there's no flushing of anything).

Synchronizing on "this" makes the code more understandable, more correct, and more performant (since constant features will never be forgotten about).

Reply:
Again, no need to synchronized on `this`. No synchronization is required.

Java Memory Model enforces memory fence. See 
http://gee.cs.oswego.edu/dl/jmm/cookbook.html
```
EnterLoad is needed on entry to any synchronized block/method that performs a load. It is the same as LoadLoad unless an atomic instruction is used in MonitorEnter and itself provides a barrier with at least the properties of LoadLoad, in which case it is a no-op.
StoreExit is needed on exit of any synchronized block/method that performs a store. It is the same as StoreStore unless an atomic instruction is used in MonitorExit and itself provides a barrier with at least the properties of StoreStore, in which case it is a no-op.
ExitEnter is the same as StoreLoad unless atomic instructions are used in MonitorExit and/or MonitorEnter and at least one of these provide a barrier with at least the properties of StoreLoad, in which case it is a no-op.
```

Reply:
> this.constFeatures' value can change

It's accepted.

> Synchronization isn't a global memory barrier

No, Java `synchronized` enforces memory barrier.